compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/InMemoryCacheWithEviction.kt

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2010-2022 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.incremental.classpathDiff
+
+import org.jetbrains.kotlin.utils.ThreadSafe
+import java.lang.ref.SoftReference
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+import javax.annotation.concurrent.GuardedBy
+
+/**
+ * In-memory cache that maintains either strong or soft references to cache entries' values so that it adapts to memory availability (see
+ * [strongCache] and [softCache]'s kdoc).
+ *
+ * This cache supports 2 types of cache eviction:
+ *   - Least recently used: See [maxTimePeriods]'s kdoc
+ *   - Memory usage limit: See [maxMemoryUsageRatio]'s kdoc
+ */
+@ThreadSafe
+class InMemoryCacheWithEviction<KEY, VALUE>(
+
+    /**
+     * Cache entries that were not used within [maxTimePeriods] will be made available for garbage collection (they will be moved from
+     * [strongCache] to [softCache]).
+     *
+     * The time period starts from 0 and will increment by 1 whenever [newTimePeriod] is called.
+     */
+    private val maxTimePeriods: Int,
+
+    /**
+     * If [memoryUsageRatio] > [maxMemoryUsageRatio], all cache entries will be made available for garbage collection (all cache entries in
+     * [strongCache] will be moved to [softCache]).
+     */
+    private val maxMemoryUsageRatio: Double
+) {
+
+    /** The current time period, which starts from 0 and will increment by 1 whenever [newTimePeriod] is called. */
+    private val currentTimePeriod = AtomicInteger(0)
+
+    /**
+     * Contains recently-used cache entries. We keep strong references to the values so that they can always be used until they are moved to
+     * [softCache], where they can be used for some more time until being garbage collected.
+     */
+    @GuardedBy("lock")
+    private val strongCache = mutableMapOf<KEY, VALUE>()
+
+    /** Contains older cache entries. We keep [SoftReference]s to the values so that they can be used until being garbage collected. */
+    @GuardedBy("lock")
+    private val softCache = mutableMapOf<KEY, SoftReference<VALUE>>()
+
+    /** Maps each key in [strongCache] to the most recent time period when it is used. */
+    @GuardedBy("lock")
+    private val cacheEntriesLastUsed = mutableMapOf<KEY, Int>()
```

Comment:
WDYT about having a single ConcurrentHashMap to contain KEY -> VALUE entries, and each entry is just a wrapper that has:
- a strong reference to cached value
- a weak ref to cached value
- last used timestamp

it may simplify things.

Also, can you please add tests for this cache (you probably need to make memory usage injectable).

Reply:
That's a great idea! It does simplify this class greatly (even though there is still heavy use of synchronization that I can't manage to avoid).

One tiny issue is that cache eviction will be slightly slower: with the previous approach, we just looked at the strong references, whereas with the new approach, we're going through all entries including both strong and soft references.

I added a unit test too.